File: frontend/src/utils/etlSourceScenarios.ts
/**
 * Supported tenant ETL source layouts (product scenarios).
 */
import type { TenantEtlSource } from '../services/etlSourcesService'

export type EtlScenarioId =
  | 'mysql_only'
  | 'one_mysql_multi_postgres'
  | 'postgres_only_multi'
  | 'postgres_only_single'
  | 'mixed_other'
  | 'legacy'
  | 'empty'

export interface EtlScenarioDefinition {
  id: EtlScenarioId
  title: string
  description: string
  /** Product scenario A/B/C — shown in the UI checklist */
  productScenario?: 'A' | 'B' | 'C'
}

export const ETL_PRODUCT_SCENARIOS: EtlScenarioDefinition[] = [
  {
    id: 'mysql_only',
    productScenario: 'A',
    title: 'MySQL only',
    description: 'One or more MySQL sources; no PostgreSQL.',
  },
  {
    id: 'one_mysql_multi_postgres',
    productScenario: 'B',
    title: '1 MySQL + 2+ PostgreSQL',
    description: 'Exactly one MySQL source and two or more PostgreSQL sources.',
  },
  {
    id: 'postgres_only_multi',
    productScenario: 'C',
    title: 'PostgreSQL only (2+)',
    description: 'Two or more PostgreSQL sources; no MySQL.',
  },
]

export interface EtlScenarioStatus {
  activeScenario: EtlScenarioId
  mysqlCount: number
  postgresCount: number
  totalRegistered: number
  hasPrimary: boolean
  matchesProductScenario: boolean
  matchedProductScenario?: 'A' | 'B' | 'C'
  hint?: string
}

export function registeredSources(sources: TenantEtlSource[]): TenantEtlSource[] {
  return sources.filter(s => s.id > 0)
}

export function analyzeEtlSources(sources: TenantEtlSource[]): EtlScenarioStatus {
  const registered = registeredSources(sources)
  const mysqlCount = registered.filter(s => s.source_type === 'mysql').length
  const postgresCount = registered.filter(s => s.source_type === 'postgres').length
  const hasPrimary = registered.some(s => s.is_primary)

  if (registered.length === 0) {
    if (sources.length > 0) {
      return {
        activeScenario: 'legacy',
        mysqlCount: 0,
        postgresCount: 0,
        totalRegistered: 0,
        hasPrimary: false,
        matchesProductScenario: false,
        hint: 'Using legacy tenant registry. Add explicit sources below.',
      }
    }
    return {
      activeScenario: 'empty',
      mysqlCount: 0,
      postgresCount: 0,
      totalRegistered: 0,
      hasPrimary: false,
      matchesProductScenario: false,
      hint: 'Add at least one source database.',
    }
  }

  let activeScenario: EtlScenarioId
  let matchedProductScenario: 'A' | 'B' | 'C' | undefined
  let matchesProductScenario = false

  if (mysqlCount >= 1 && postgresCount === 0) {
    activeScenario = 'mysql_only'
    matchedProductScenario = 'A'
    matchesProductScenario = true
  } else if (mysqlCount === 1 && postgresCount >= 2) {
    activeScenario = 'one_mysql_multi_postgres'
    matchedProductScenario = 'B'
    matchesProductScenario = true
  } else if (mysqlCount === 0 && postgresCount >= 2) {
    activeScenario = 'postgres_only_multi'
    matchedProductScenario = 'C'
    matchesProductScenario = true
  } else if (mysqlCount === 0 && postgresCount === 1) {
    activeScenario = 'postgres_only_single'
    matchesProductScenario = false
  } else {
    activeScenario = 'mixed_other'
    matchesProductScenario = false
  }

  let hint: string | undefined
  if (!hasPrimary && registered.length > 1) {
    hint = 'Mark one source as Primary so standard stg_* tables have a clear owner.'
  } else if (activeScenario === 'mixed_other') {
    if (mysqlCount > 1 && postgresCount >= 1) {
      hint =
        'You have multiple MySQL sources. Scenario B expects exactly one MySQL; ETL will still run all sources.'
    } else if (mysqlCount === 1 && postgresCount === 1) {
      hint =
        'Scenario B expects 1 MySQL + 2+ PostgreSQL. Add another PostgreSQL source or use Scenario A/C layout.'
    }
  }

  return {
    activeScenario,
    mysqlCount,
    postgresCount,
    totalRegistered: registered.length,
    hasPrimary,
    matchesProductScenario,
    matchedProductScenario,
    hint,
  }
}

File: backend/app/services/etl_source_scenarios.py
"""Classify tenant ETL source layouts (product scenarios A/B/C)."""
from __future__ import annotations

from typing import Any, Literal, Optional

from app.services.tenant_etl_sources import TenantEtlSourceRow

ScenarioId = Literal[
    "mysql_only",
    "one_mysql_multi_postgres",
    "postgres_only_multi",
    "postgres_only_single",
    "mixed_other",
    "legacy",
    "empty",
]


def _registered(rows: list[TenantEtlSourceRow]) -> list[TenantEtlSourceRow]:
    return [r for r in rows if r.id != 0]


def classify_etl_sources(rows: list[TenantEtlSourceRow]) -> dict[str, Any]:
    """Return counts and which product scenario (if any) matches."""
    registered = _registered(rows)
    mysql = [r for r in registered if r.source_type == "mysql"]
    pg = [r for r in registered if r.source_type == "postgres"]
    mysql_count = len(mysql)
    postgres_count = len(pg)

    if not registered:
        if rows:
            return {
                "active_scenario": "legacy",
                "mysql_count": 0,
                "postgres_count": 0,
                "total_registered": 0,
                "has_primary": False,
                "matches_product_scenario": False,
                "product_scenario": None,
                "etl_ready": True,
            }
        return {
            "active_scenario": "empty",
            "mysql_count": 0,
            "postgres_count": 0,
            "total_registered": 0,
            "has_primary": False,
            "matches_product_scenario": False,
            "product_scenario": None,
            "etl_ready": False,
        }

    active: ScenarioId
    product: Optional[str] = None
    matches = False

    if mysql_count >= 1 and postgres_count == 0:
        active, product, matches = "mysql_only", "A", True
    elif mysql_count == 1 and postgres_count >= 2:
        active, product, matches = "one_mysql_multi_postgres", "B", True
    elif mysql_count == 0 and postgres_count >= 2:
        active, product, matches = "postgres_only_multi", "C", True
    elif mysql_count == 0 and postgres_count == 1:
        active, product, matches = "postgres_only_single", None, False
    else:
        active, product, matches = "mixed_other", None, False

    primary_key = next((r.source_key for r in registered if r.is_primary), None)

    return {
        "active_scenario": active,
        "mysql_count": mysql_count,
        "postgres_count": postgres_count,
        "total_registered": len(registered),
        "has_primary": primary_key is not None,
        "primary_source_key": primary_key,
        "matches_product_scenario": matches,
        "product_scenario": product,
        "etl_ready": True,
        "sources": [
            {
                "source_key": r.source_key,
                "display_name": r.display_name,
                "source_type": r.source_type,
                "source_schema": r.source_schema,
                "is_primary": r.is_primary,
                "staging_suffix": r.staging_suffix or "(primary stg_*)",
            }
            for r in registered
        ],
    }
